doozerlib/assembly.py

```diff
@@ -0,0 +1,138 @@
+import typing
+import copy
+
+from doozerlib.model import Model
+
+
+def merger(a, b):
+    """
+    Merges two, potentially deep, objects into a new one and returns the result.
+    Conceptually, 'a' is layered over 'b' and is dominant when
+    necessary. The output is 'c'.
+    1. if 'a' specifies a primitive value, regardless of depth, 'c' will contain that value.
+    2. if a key in 'a' specifies a list and 'b' has the same key/list, a's list will be appended to b's for c's list.
+       Duplicates entries will be removed and primitive (str, int, ..) lists will be returned in sorted order).
+    3. if a key ending with '!' in 'a' specifies a value, c's key-! will be to set that value exactly.
+    4. if a key ending with a '?' in 'a' specifies a value, c's key-? will be set to that value is 'c' does not contain the key.
+    """
+
+    if type(a) in [bool, int, float, str, bytes, type(None)]:
+        return a
+
+    c = copy.deepcopy(b)
+
+    if type(a) is list:
+        if type(c) is not list:
+            return a
+        for entry in a:
+            if entry not in c:  # do not include duplicates
+                c.append(entry)
+
+        if c and type(c[0]) in [str, int, float]:
+            return sorted(c)
+        return c
+
+    if type(a) is dict:
+        if type(c) is not dict:
+            return a
+        for k, v in a.items():
+            if k.endswith('!'):  # full dominant key
+                k = k[:-1]
+                c[k] = v
+            elif k.endswith('?'):  # default value key
+                k = k[:-1]
+                if k not in c:
+                    c[k] = v
+            else:
+                if k in c:
+                    c[k] = merger(a[k], c[k])
+                else:
+                    c[k] = a[k]
+        return c
+
+    raise IOError(f'Unexpected value type: {type(a)}: {a}')
```

Comment:
Seems a `TypeError` is more appropriate. 

```suggestion
    raise TypeError(f'Unexpected value type: {type(a)}: {a}')
```

Reply:
ack. Will be in next push.